src/test/clojure/clara/generative_tests/generators.clj

```diff
@@ -0,0 +1,102 @@
+                                                    :retract (apply retract session (:facts op))
+                                                    :fire (fire-rules session))]
+                                  new-session))
+                              session ops)]))
+
+(defn ^:private retract-before-insertion?
+  "Given a sequence of operations, determine if the number of retractions of any fact exceeds the number
+  of times has been inserted at any time."
+  [ops]
+  (let [alter-fact-count (fn [alter-fn]
+                           (fn [fc facts]
+                             (reduce (fn [updated-fc f]
+                                       (update fc f alter-fn))
+                                     fc facts)))
+        inc-fact-count (alter-fact-count (fnil inc 0))
+        dec-fact-count (alter-fact-count (fnil dec 0))
+
+        any-count-negative? (fn [fc]
+                              (boolean (some neg? (vals fc))))]
+    
+    (= ::premature-retract (reduce (fn [fact-count op]
+                                     (let [new-count (condp = (:type op)
+                                                       :insert (inc-fact-count fact-count (:facts op))
+                                                       :retract (dec-fact-count fact-count (:facts op))
+                                                       fact-count)]
+                                       (if (any-count-negative? new-count)
+                                         (reduced ::premature-retract)
+                                         new-count)))
+                                   {}
+                                   ops))))
+
+(defn ^:private ops->add-insert-retract
+  "Takes an operations sequence and returns a sequence of sequences of operations
+  where every possible combination of inserting and retracting each fact inserted
+  in the parent sequence up to the number of times set by dup-level.  Note that one
+  of the possibilities returned will be the original operations sequence.  This is not
+  special-cased, but a reflection of the fact that every combination of adding insert/retract pairs
+  from 0 to dup-level is in the possibilities returned.  The number of possibilities returned
+  will explode rapidly for large dup-level values."
+  [ops dup-level]
+  (let [ops->extra (fn [ops]
+                     (map (fn [op]
+                            (when (= (:type op)
+                                     :insert)
+                              [{:type :insert
+                                :facts (:facts op)}
+                               {:type :retract
+                                :facts (:facts op)}]))
+                          ops))
+        extras-with-dups (apply concat (repeat dup-level
+                                               (ops->extra ops)))
+
+        extra-subsets (combo/subsets extras-with-dups)]
+
+    (map (fn [extras]
+           (into ops (apply concat extras)))
```

Comment:
nit: `(into ops cat extras)` would be the same thing (in behavior, not performance)